File: app/src/main/java/com/zebrostudio/lowpolyrx/MainActivity.java
package com.zebrostudio.lowpolyrx;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.app.AppCompatActivity;
import com.afollestad.materialdialogs.MaterialDialog;
import com.zebrostudio.imagecomparisonview.ImageComparisonView;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class MainActivity extends AppCompatActivity {

  private Disposable disposable;

  @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    final MaterialDialog lowpolyWaitLoader = new MaterialDialog.Builder(this)
        .widgetColor(colorRes(R.color.white))
        .contentColor(colorRes(R.color.white))
        .content(stringRes(R.string.lowpoly_wait_loader_message))
        .backgroundColor(colorRes(R.color.colorPrimary))
        .progress(true, 0)
        .progressIndeterminateStyle(false)
        .cancelable(false)
        .build();

    lowpolyWaitLoader.show();

    disposable = generateLowpolyFromResource()
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Consumer<Bitmap>() {
          @Override public void accept(Bitmap bitmap) {
            ImageComparisonView imageComparisonView = findViewById(R.id.imageComparisonView);
            imageComparisonView.setImages(R.mipmap.captain, bitmap);
            lowpolyWaitLoader.dismiss();
          }
        }, new Consumer<Throwable>() {
          @Override public void accept(Throwable throwable) {
            lowpolyWaitLoader.dismiss();
          }
        });
  }

  @Override
  protected void onDestroy() {
    if (!disposable.isDisposed()){
      disposable.dispose();
    }
    super.onDestroy();
  }

  private Single<Bitmap> generateLowpolyFromResource() {
    return new LowPolyRx().getLowPolyImage(this, R.mipmap.captain);
  }

  private int colorRes(@ColorRes int color) {
    return this.getResources().getColor(color);
  }

  private String stringRes(@StringRes int string) {
    return this.getString(string);
  }
}

File: app/src/main/java/com/zebrostudio/lowpolyrx/MainActivity.kt
package com.zebrostudio.lowpolyrx

import android.arch.lifecycle.Lifecycle
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Environment
import android.support.v7.app.AppCompatActivity
import com.afollestad.materialdialogs.MaterialDialog
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider
import com.uber.autodispose.autoDisposable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_main.imageComparisonView
import java.io.File

class MainActivity : AppCompatActivity() {

  private var lowpolyWaitLoader: MaterialDialog? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    lowpolyWaitLoader = MaterialDialog.Builder(this)
        .widgetColor(colorRes(R.color.white))
        .contentColor(colorRes(R.color.white))
        .content(getString(R.string.lowpoly_wait_loader_message))
        .backgroundColor(colorRes(R.color.colorPrimary))
        .progress(true, 0)
        .progressIndeterminateStyle(false)
        .cancelable(false)
        .build()

    lowpolyWaitLoader?.show()

    generateLowpoly()
        .observeOn(AndroidSchedulers.mainThread())
        .autoDisposable(AndroidLifecycleScopeProvider.from(this, Lifecycle.Event.ON_DESTROY))
        .subscribe({
          imageComparisonView.setImages(R.mipmap.sample_large, it)
          val file = File(Environment.getExternalStorageDirectory(), "Sample6.jpg")
          file.createNewFile()
          file.writeBitmap(it, Bitmap.CompressFormat.JPEG, 100)
          lowpolyWaitLoader?.dismiss()
        }, {
          lowpolyWaitLoader?.dismiss()
        })

  }

  private fun generateLowpoly(): Single<Bitmap> {
    return LowPolyRx().getLowPolyImage(this, R.mipmap.sample_large)
  }

}

private fun File.writeBitmap(
  bitmap: Bitmap,
  format: Bitmap.CompressFormat,
  quality: Int
) {
  outputStream().use { out ->
    bitmap.compress(format, quality, out)
    out.flush()
  }
}
